src/ptvsd/common/launcher.py

```diff
@@ -0,0 +1,76 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See LICENSE in the project root
+# for license information.
+
+from __future__ import absolute_import, print_function, unicode_literals
+
+
+import subprocess
+import sys
+
+
+_wait_on_normal_exit = False
+_wait_on_abnormal_exit = False
+
+
+HELP = """Usage: launcher [--wait-on-normal] [--wait-on-abnormal] <args>
+python launcher.py --wait-on-normal --wait-on-abnormal myscript.py -a 123      # script with args
+python launcher.py --wait-on-normal --wait-on-abnormal -m mymodule -a 123      # module with args
+python launcher.py --wait-on-normal --wait-on-abnormal -c "<code>"             # run code
+"""
+
+
+def main(argv=sys.argv):
+    try:
+        process_args = list(parse(argv))
+    except Exception as ex:
+        print(HELP + "\nError: " + str(ex), file=sys.stderr)
+        sys.exit(2)
+
+    p = subprocess.Popen(args=process_args, executable=sys.executable)
+    exit_code = p.wait()
+
+    if _wait_on_normal_exit and exit_code == 0:
+        _wait_for_user()
+    elif _wait_on_abnormal_exit and exit_code != 0:
+        _wait_for_user()
+
+    sys.exit(exit_code)
+
+
+def _wait_for_user():
+    if sys.__stdout__ and sys.__stdin__:
+        try:
+            import msvcrt
+        except ImportError:
+            sys.__stdout__.write("Press Enter to continue . . . ")
+            sys.__stdout__.flush()
+            sys.__stdin__.read(1)
+        else:
+            sys.__stdout__.write("Press any key to continue . . . ")
+            sys.__stdout__.flush()
+            msvcrt.getch()
+
+
+def parse_arg(arg):
+    if arg == "--wait-on-normal":
+        global _wait_on_normal_exit
+        _wait_on_normal_exit = True
+    elif arg == "--wait-on-abnormal":
+        global _wait_on_abnormal_exit
+        _wait_on_abnormal_exit = True
+    else:
+        return False
+    return True
+
+
+def parse(argv):
```

Comment:
This will break if the user does:
```
python launcher.py myscript.py --wait-on-normal
```
(i.e. if it's supposed to be a switch for their own script, that just happens to use the same name)

To work in this case, it needs to do something similar to what arg parsing code in `ptvsd/server/__main__.py` does today. So basically parse them, separate out our args like `--wait` from Python args like `-X`, and stop once it sees the filename, or `-m`, or `-c`.

But this is complicated. Since this script isn't meant to be run by the users, I think we can do it much easier by requiring an explicit separator, e.g. `--`. So the adapter would always invoke it like so:
```
python launcher.py --wait-on-normal -- myscript.py --wait-on-normal
```
and then it just needs to parse up to `--` (and complain about any invalid switches in that part), and hand everything after `--` to Python as is.